storage/imx6ull-flashnor/flashnor-qspi.c

```diff
@@ -0,0 +1,303 @@
+/*
+ * Phoenix-RTOS
+ *
+ * i.MX 6ULL QSPI NOR flash driver
+ *
+ * Copyright 2023 Phoenix Systems
+ * Author: Hubert Badocha
+ *
+ * This file is part of Phoenix-RTOS.
+ *
+ * %LICENSE%
+ */
+
+#include <errno.h>
+#include <string.h>
+#include <unistd.h>
+#include <sys/threads.h>
+#include <sys/types.h>
+#include <qspi.h>
+#include "flashnor-qspi.h"
+
+
+#define CMD_WREN       0x06 /* Write Enable */
+#define CMD_RDSR       0x05 /* Read Status Register */
+#define CMD_4QWRITE    0x34 /* 4-byte Quad I/O Write */
+#define CMD_QENABLE    0x35 /* Enable Quad I/O */
+#define CMD_4ERASE4KB  0x21 /*  4-byte  Sector Erase (4 KB) */
+#define CMD_JEDEC      0x9f /* JEDEC ID */
+#define CMD_4QFREAD_IO 0xec /*  4-byte Quad Fast Read I/O */
+
+#define FLASH_PAGE_SIZE 256
+
+enum {
+	lut_seq_jedec = 0,
+	lut_seq_write_enable,
+	lut_seq_write,
+	lut_seq_erase,
+	lut_seq_read,
+	lut_seq_read_status,
+	lut_seq_quad_io,
+};
+
+typedef struct {
+	const char *name;       /* Chip name */
+	unsigned char jedec[3]; /* Chip JEDEC ID */
+	unsigned int sectorsz;  /* Chip sector size in bytes */
+	unsigned int flashsz;   /* Chip storage size in bytes */
+} chip_t;
+
+/* Supported NOR flash chips */
+static const chip_t chips[] = {
+	{ "Micron MT25QL256ABA(256Mb, 3V)", { 0x20, 0xba, 0x19 }, 4 * 1024, 32 * 1024 * 1024 },
+	/* Not tested.
+
+	{ "Micron MT25QL256ABA(1Gb, 3V)", { 0x20, 0xba, 0x21 }, 64 * 1024, 1024 * 1024 * 1024 },
+	{ "Micron MT25QL256ABA(512Mb, 3V)", { 0x20, 0xba, 0x20 }, 64 * 1024, 512 * 1024 * 1024 },
+	{ "Micron MT25QL256ABA(128Mb, 3V)", { 0x20, 0xba, 0x18 }, 64 * 1024, 128 * 1024 * 1024 },
+	{ "Micron MT25QL256ABA(16Mb, 3V)", { 0x20, 0xba, 0x17 }, 64 * 1024, 16 * 1024 * 1024 },
+	{ "Micron MT25QL256ABA(1Gb, 1.8V)", { 0x20, 0xbb, 0x21 }, 64 * 1024, 1024 * 1024 * 1024 },
+	{ "Micron MT25QL256ABA(512Mb, 1.8V)", { 0x20, 0xbb, 0x20 }, 64 * 1024, 512 * 1024 * 1024 },
+	{ "Micron MT25QL256ABA(256Mb, 1.8V)", { 0x20, 0xbb, 0x19 }, 64 * 1024, 256 * 1024 * 1024 },
+	{ "Micron MT25QL256ABA(128Mb, 1.8V)", { 0x20, 0xbb, 0x18 }, 64 * 1024, 128 * 1024 * 1024 },
+	{ "Micron MT25QL256ABA(16Mb, 1.8V)", { 0x20, 0xbb, 0x17 }, 64 * 1024, 16 * 1024 * 1024 },
+	*/
+};
+
+struct {
+	handle_t lock;
+	int init;
+} flashnor_common;
+
+
+static int _flashnor_qspiWriteEnable(qspi_dev_t dev)
+{
+	return _qspi_readBusy(dev, lut_seq_write_enable, 0, NULL, 0);
+}
+
+
+static int _flashnor_qspiWaitBusy(qspi_dev_t dev)
+{
+	uint8_t status;
+	int err, max_iter;
+	unsigned int sleep = 100;
+
+
+	for (max_iter = 1000; max_iter > 0; max_iter--) {
+		err = _qspi_readBusy(dev, lut_seq_read_status, 0, &status, 1);
+		if (err < 0) {
+			return err;
+		}
+
+		if ((status & 1) == 0) {
+			return EOK;
+		}
+
+		usleep(sleep);
+
+		if (sleep < 100000) {
+			sleep <<= 1;
+		}
+	}
+	return -ETIME;
+}
+
+
+ssize_t flashnor_qspiRead(qspi_dev_t dev, unsigned int addr, void *buff, size_t bufflen)
+{
+	size_t size, len;
+	int err;
+
+	err = mutexLock(flashnor_common.lock);
```

Comment:
It's ok to ignore retval of `mutexLock`/`mutexUnlock` (only possible error: `-EINVAL` on non-existing mutex, we can guarantee mutex being valid by checking `mutexCreate` reval instead). Do `(void)` cast to satisfy MISRA.